Change creation and validation of OverrideCycles

diitaz93 opened the pull request:
## Description
Closes #3404
Changes the creation and validation of sample sheets so that they allow for demultiplexing without indexes

### Added

- Statement in the override cycles validator that passes validation if there is no index1 and all positions of index1 are ignored
- Statement in the sample index class making override cycles ignore all index positions if no indexes are provided
- Staemnent in the sample index class making barcode mismatches 1 equal to 'na' if there is no index1
- Fixtures and tests

### Changed

- Allow the barcode mismatches 1 to be str ('na')


### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b patch-override-cycles -a
    ```

### How to test

- [x] Create sample sheet `cg -l DEBUG demultiplex samplesheet create 161206_ST-E00201_0170_BH57JWALXX`
- [x] Validate the sample sheet with `cg -l DEBUG demultiplex samplesheet validate SampleSheet.csv`
- [x] Dry-run demultiplex
- [x] Demultiplex using dragen
- [x] Post-process flow cell

### Expected test outcome

Check that ...
- [x] Sample sheet is created successfully, there are no indexes, override cycles for index is "n8" and barcode mismatches1 is "na"
- [x] Sample sheet passes validation
- [x] The sample sheet can be used in demultiplexing
- [x] The post-processing works

## Review

- [x] Tests executed by SD
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Deployed to stage:
```shell

```
- [ ] Deployed to production:
```shell

```

Karl-Svard commented on cg/apps/demultiplex/sample_sheet/sample_models.py:
```diff
@@ -85,6 +87,9 @@ def _update_barcode_mismatches_1(
         """Assign zero to barcode_mismatches_1 if the hamming distance between self.index
         and the index1 of any sample in the lane is below the minimum threshold."""
+        if not self.index:
+            self.barcode_mismatches_1 = "na"
```
Is "na" an allowed value? From the documentation online it looks like it only accepts 0, 1 or 2. https://support.illumina.com/help/BS_App_BCL_Convert/v110_1000000130493/Content/SW/Apps/SampleSheetSettings_swBCL_swBS_appBCL.htm

diitaz93 replied:
Resolved in person

diitaz93 commented:
## Tests on stage
Create sample sheet:
```shell
$ cg -l DEBUG demultiplex samplesheet create 161206_ST-E00201_0170_BH57JWALXX
Running cg demultiplex.
Getting a valid sample sheet for flow cell 161206_ST-E00201_0170_BH57JWALXX
Instantiating sample sheet API
Instantiating housekeeper api
Initializing Store
Instantiating lims api
Called undefined __fields__ on HousekeeperAPI, please wrap
Set dry run to False
Set force to False
Instantiating IlluminaRunDirectoryData with path /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX
Set sequencing run id to BH57JWALXX
Fetching and validating sample sheet for 161206_ST-E00201_0170_BH57JWALXX from Housekeeper
Fetch latest version from bundle H57JWALXX
Fetching files with tags in [H57JWALXX,samplesheet]
Fetching files from version 75433
Validating sample sheet
Validating that the sample sheet has all the necessary sections
Sample sheet does not have all the necessary sections
Sample sheet from Housekeeper is not correctly formatted or does not exist, trying sample sheet in sequencing directory
Sample sheet with path /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/SampleSheet.csv does not exist
Sample sheet from sequencing directory is not correctly formatted or does not exist, creating new sample sheet
Fetching samples from lims for flowcell H57JWALXX
Starting new HTTPS connection (1): cg-lims-stage.sys.scilifelab.se
...
Constructing sample sheet for the hiseqx flow cell H57JWALXX
Order samples by lane
Updating barcode mismatch values for samples in lane 1
Run is single-indexed, skipping barcode mismatch update for index 2
Updating barcode mismatch values for samples in lane 2
Run is single-indexed, skipping barcode mismatch update for index 2
Updating barcode mismatch values for samples in lane 3
Run is single-indexed, skipping barcode mismatch update for index 2
Updating barcode mismatch values for samples in lane 4
Run is single-indexed, skipping barcode mismatch update for index 2
Updating barcode mismatch values for samples in lane 5
Run is single-indexed, skipping barcode mismatch update for index 2
Updating barcode mismatch values for samples in lane 6
Run is single-indexed, skipping barcode mismatch update for index 2
Updating barcode mismatch values for samples in lane 7
Run is single-indexed, skipping barcode mismatch update for index 2
Updating barcode mismatch values for samples in lane 8
Run is single-indexed, skipping barcode mismatch update for index 2
Creating sample sheet content
Validating sample sheet
Validating that the sample sheet has all the necessary sections
Looking for index settings in the sample sheet
Found index settings: NoReverseComplements
Looking for read and index run cycles in the sample sheet
Validating samples
Order samples by lane
Validate that samples are unique in lane: 1
Validate that samples are unique in lane: 2
Validate that samples are unique in lane: 3
Validate that samples are unique in lane: 4
Validate that samples are unique in lane: 5
Validate that samples are unique in lane: 6
Validate that samples are unique in lane: 7
Validate that samples are unique in lane: 8
Validating override cycles for all samples
Samplesheet passed validation
Writing sample sheet to /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/SampleSheet.csv
Fetch latest version from bundle H57JWALXX
Fetching files with tags in [H57JWALXX,samplesheet]
Fetching files from version 75433
Fetching file with path /home/proj/stage/housekeeper-bundles/H57JWALXX/2022-01-18/SampleSheet.csv
Fetching file with path H57JWALXX/2022-01-18/SampleSheet.csv
Return file: 6238756
Deleting file /home/proj/stage/housekeeper-bundles/H57JWALXX/2022-01-18/SampleSheet.csv from disc
Deleting file 6238756 from housekeeper
Adding sample sheet to Housekeeper
Fetching bundle with name: H57JWALXX
Bundle with name H57JWALXX already exists
Fetch latest version from bundle H57JWALXX
Found Housekeeper version object for H57JWALXX: <housekeeper.store.models.Version object at 0x7f7932305c50>
Fetch latest version from bundle H57JWALXX
Fetching tag with name: samplesheet
Fetching tag with name: H57JWALXX
Fetching tag with name: samplesheet
Fetching tag with name: H57JWALXX
Created new bundle version dir: /home/proj/stage/housekeeper-bundles/H57JWALXX/2022-01-18
Linked file: /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/SampleSheet.csv -> /home/proj/stage/housekeeper-bundles/H57JWALXX/2022-01-18/SampleSheet.csv
File added to Housekeeper bundle H57JWALXX
$
```

islean commented on cg/apps/demultiplex/sample_sheet/override_cycles_validator.py:
```diff
@@ -104,6 +104,11 @@ def _validate_index1_cycles(self) -> None:
             OverrideCyclesError if the index 1 cycle is not valid.
         """
         index1_cycle: str = self.sample_cycles[1]
+        if (
+            len(self.sample[SampleSheetBCLConvertSections.Data.INDEX_1]) == 0
+            and index1_cycle == f"N{self.run_index1_cycles}"
+        ):
```
Can we add a little utils method for this check?

diitaz93 commented:
## Tests on stage
Validate sample sheet
```shell
$ cg -l DEBUG demultiplex samplesheet validate SampleSheet.csv
Running cg demultiplex.
Validating SampleSheet.csv sample sheet
Instantiating sample sheet API
Instantiating housekeeper api
Initializing Store
Instantiating lims api
Called undefined __fields__ on HousekeeperAPI, please wrap
Validating sample sheet
Validating that the sample sheet has all the necessary sections
Looking for index settings in the sample sheet
Found index settings: NoReverseComplements
Looking for read and index run cycles in the sample sheet
Validating samples
Order samples by lane
Validate that samples are unique in lane: 1
Validate that samples are unique in lane: 2
Validate that samples are unique in lane: 3
Validate that samples are unique in lane: 4
Validate that samples are unique in lane: 5
Validate that samples are unique in lane: 6
Validate that samples are unique in lane: 7
Validate that samples are unique in lane: 8
Validating override cycles for all samples
Samplesheet passed validation
Sample sheet passed validation
$
```

diitaz93 commented:
## Test on stage
Demultiplexing
```shell
$ cg -l DEBUG demultiplex sequencing-run 161206_ST-E00201_0170_BH57JWALXX --dry-run
Running cg demultiplex.
Starting demultiplexing of sequencing run 161206_ST-E00201_0170_BH57JWALXX
Instantiating sample sheet API
Instantiating housekeeper api
Initializing Store
Instantiating lims api
Called undefined __fields__ on HousekeeperAPI, please wrap
Instantiating demultiplexing api
Called undefined __fields__ on HousekeeperAPI, please wrap
Initialising Process with binary: sbatch
Use base call ['sbatch']
Set environment to stage
DemultiplexingAPI: Set dry run to True
SlurmAPI: Set dry run to True
setting flow cell id to 161206_ST-E00201_0170_BH57JWALXX
setting demultiplexed runs dir to /home/proj/stage/sequencing_data/illumina/demultiplexed-runs
Instantiating IlluminaRunDirectoryData with path /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX
Set sequencing run id to BH57JWALXX
Check if demultiplexing is possible for H57JWALXX
Check if sequencing run is ready for downstream processing
Check if sequencing is done
Sequence is done for sequencing run H57JWALXX
Check if copy of data from sequence instrument is ready
All data has been transferred for sequencing run H57JWALXX
Sequencing run H57JWALXX is ready for downstream processing
Check if sample sheet exists
Fetch latest version from bundle H57JWALXX
Fetching files with tags in [H57JWALXX,samplesheet]
Fetching files from version 75433
Validating sample sheet
Validating that the sample sheet has all the necessary sections
Looking for index settings in the sample sheet
Found index settings: NoReverseComplements
Looking for read and index run cycles in the sample sheet
Validating samples
Order samples by lane
Validate that samples are unique in lane: 1
Validate that samples are unique in lane: 2
Validate that samples are unique in lane: 3
Validate that samples are unique in lane: 4
Validate that samples are unique in lane: 5
Validate that samples are unique in lane: 6
Validate that samples are unique in lane: 7
Validate that samples are unique in lane: 8
Validating override cycles for all samples
Samplesheet passed validation
Would have started demultiplexing 161206_ST-E00201_0170_BH57JWALXX
$
```
DRAGEN
```shell
Sample sheet being processed by common lib? Yes
SampleSheet Settings:
  FastqCompressionFormat = gzip
  SoftwareVersion = 4.1.7

shared-thread-linux-native-asio output is disabled
Generating run log at /var/log/dragen/dragen_run_1727707018006_206910.log

Command Line: dragen --bcl-conversion-only true --bcl-input-directory /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX --output-directory /home/proj/stage/sequencing_data/illumina/demultiplexed-runs/161206_ST-E00201_0170_BH57JWALXX --force --sample-sheet /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/SampleSheet.csv
DRAGEN Host Software Version 4.1.7
DRAGEN Daemon Software Version 07.031.650.4.1.7 and Bio-IT Processor Version 0x18101306

AutoDetected reference: UNKNOWN
PR assigned board: -Z0

==================================================================
Downloading ZIP/UNZIP HW bitstream to board# 0 - do not interrupt
==================================================================
Info: Requested image is already loaded.
Info: Found 10 engines for AFU zip
# CPU hw threads available: 48
Parallel Tiles: 4. Threads Per Tile: 12
HW compressors: 10
HW decompressors: 0
SW compressors: 0
SW decompressors: 24
SW FASTQ compression level: 1
  Sep 30 2024  16:36:58:744:341  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2063 size = 30248 stream = 10
  Sep 30 2024  16:36:58:765:695  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2064 size = 30248 stream = 11
  Sep 30 2024  16:36:58:789:641  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2065 size = 30248 stream = 12
  Sep 30 2024  16:36:58:809:742  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2066 size = 30248 stream = 13
  Sep 30 2024  16:36:58:819:266  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2067 size = 30248 stream = 14
  Sep 30 2024  16:36:58:828:024  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2068 size = 30248 stream = 15
  Sep 30 2024  16:36:58:853:433  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2069 size = 30248 stream = 0
  Sep 30 2024  16:36:58:860:380  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2070 size = 30248 stream = 1
  Sep 30 2024  16:36:58:868:406  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2071 size = 30248 stream = 2
  Sep 30 2024  16:36:58:876:952  DCLC-0268  (206910) || client init: /dev/shm/dragend-client-afuif.cl353_a239_2072 size = 30248 stream = 3
Hardware compressor 0 finished. Processed 232510 buffers
Hardware compressor 1 finished. Processed 232242 buffers
Hardware compressor 2 finished. Processed 232593 buffers
Hardware compressor 3 finished. Processed 232750 buffers
Hardware compressor 4 finished. Processed 232196 buffers
Hardware compressor 5 finished. Processed 232243 buffers
Hardware compressor 6 finished. Processed 232659 buffers
Hardware compressor 7 finished. Processed 232618 buffers
Hardware compressor 8 finished. Processed 232647 buffers
Hardware compressor 9 finished. Processed 232488 buffers

RUN TIME                                                                         Time converting BCL to fastq                                      01:18:03.087   4683.09
RUN TIME                                                                         Total runtime                                                     01:18:03.372   4683.37
==================================================================
DRAGEN finished normally
[17:55] [hiseq.clinical@cg-dragen2:~] [base]  $
```

diitaz93 commented:
## Tests on stage
Post-processing
```shell
$ cg -l DEBUG demultiplex finish illumina-run 161206_ST-E00201_0170_BH57JWALXX
Running cg demultiplex.
Instantiating status db
Instantiating housekeeper api
Initializing Store
Post-process Illumina run 161206_ST-E00201_0170_BH57JWALXX
Instantiating IlluminaRunDirectoryData with path /home/proj/stage/sequencing_data/illumina/demultiplexed-runs/161206_ST-E00201_0170_BH57JWALXX
Set sequencing run id to BH57JWALXX
Fetch latest version from bundle H57JWALXX
Fetching files with tags in [H57JWALXX,samplesheet]
Fetching files from version 75433
Set path for Housekeeper sample sheet in run directory
Found sample sheet /home/proj/stage/housekeeper-bundles/H57JWALXX/2022-01-18/SampleSheet.csv in housekeeper.
Validating sample sheet
Validating that the sample sheet has all the necessary sections
Looking for index settings in the sample sheet
Found index settings: NoReverseComplements
Looking for read and index run cycles in the sample sheet
Validating samples
Order samples by lane
Validate that samples are unique in lane: 1
Validate that samples are unique in lane: 2
Validate that samples are unique in lane: 3
Validate that samples are unique in lane: 4
Validate that samples are unique in lane: 5
Validate that samples are unique in lane: 6
Validate that samples are unique in lane: 7
Validate that samples are unique in lane: 8
Validating override cycles for all samples
Samplesheet passed validation
Flow cell H57JWALXX has at least one sample with fastq files
Fetching files with tags in [fastq,H57JWALXX]
...
Created new bundle version dir: /home/proj/stage/housekeeper-bundles/SVE2126A30/2017-11-20
Linked file: /home/proj/stage/sequencing_data/illumina/demultiplexed-runs/161206_ST-E00201_0170_BH57JWALXX/Undetermined_S0_L008_R1_001.fastq.gz -> /home/proj/stage/housekeeper-bundles/SVE2126A30/2017-11-20/Undetermined_S0_L008_R1_001.fastq.gz
File added to Housekeeper bundle SVE2126A30
Fetch latest version from bundle H57JWALXX
Found Housekeeper version object for H57JWALXX: <housekeeper.store.models.Version object at 0x7efecc49c550>
Bundle H57JWALXX already has a file with the same name as /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/H57JWALXX_demultiplex.stderr
Added demux log file /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/H57JWALXX_demultiplex.stderr to Housekeeper.
Fetch latest version from bundle H57JWALXX
Found Housekeeper version object for H57JWALXX: <housekeeper.store.models.Version object at 0x7efecc49c990>
Bundle H57JWALXX already has a file with the same name as /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/H57JWALXX_demultiplex.stdout
Added demux log file /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/H57JWALXX_demultiplex.stdout to Housekeeper.
Fetch latest version from bundle H57JWALXX
Found Housekeeper version object for H57JWALXX: <housekeeper.store.models.Version object at 0x7efecc49e010>
Bundle H57JWALXX already has a file with the same name as /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/runParameters.xml
Added run parameters file /home/proj/stage/sequencing_data/illumina/sequencing-runs/161206_ST-E00201_0170_BH57JWALXX/runParameters.xml to H57JWALXX in Housekeeper.
[js.diazboada@hasta:161206_ST-E00201_0170_BH57JWALXX] [S_base] $
```

diitaz93 commented:
@Karl-Svard found the root cause of this issue in LIMS and patched it for the affected indexes. Re-open and inform KSV if this happens again
